### PR TITLE
Correct clustermesh identity sync kvstore backend usage (to actually use the remote)

### PR DESCRIPTION
**Pull request by raybejjani**
_This PR is easier to review commit-by-commit. There are many changes but they are mostly moving behaviour around, or passing in the backend variable instead of using the global._

Clustermesh was broken. Routing, ipcache and node sync worked but policy did not. This meant policy drops because policy maps never saw the remote identities, but ipcache did and the datapath would match correctly.

Clustermesh identity sync accidentally used the cilium-agent local etcd backend. This was because `kvstoreBackend`, the middle-layer between the identity allocators and etcd/consul, accidentally called the `pkg/kvstore` functions. These functions then used the global etcd client, which is the one talking to the cluster-local etcd.
Clustermesh identity sync also accidentally never started. This was because `CachingIdentityAllocator.WatchRemoteIdentities` ignored the etcd backend passed in and called the watch functions on the global identity allocator. This had the added effect that it the allocator added itself to itself as a remote cluster allocator, causing identity list to double print identities.

ipcache and node sync worked because they used the correct, remote, etcd backend. The wiring around allocators did the wrong thing. This has likely been the case since the abstractions around CRD identities were added.

Edit 2020-02-14
Something is broken, but I don't know what. When testing with our clustermesh tutorial (https://cilium.readthedocs.io/en/latest/gettingstarted/clustermesh/) and these steps (https://gist.github.com/raybejjani/1272bea3e9d2beb40bcf4d115768b785) I see identities propagate across clusters but the policymap never gets remote identities. I even see log lines like `level=debug msg="ToMapState: Allowed remote IDs" EndpointSelector="&LabelSelector{MatchLabels:map[string]string{k8s.io.cilium.k8s.policy.cluster: cluster2,k8s.io.kubernetes.pod.namespace: default,k8s.name: rebel-base,},MatchExpressions:[]LabelSelectorRequirement{},}" policyID="[]" subsys=policy trafficDirection=Egress` but it never uses the remote IDs :/

Edit 2020-02-17
I just don't know. I'm pretty sure the issue is something in the policy resolution and not with clustermesh directly. Someone else should try this out in case I just spent 4 days debugging a typo in a policy. 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10185)

### CODE REVIEW

**Review by jrajahalme — APPROVED** [review]
Good work!

**Comment by raybejjani** [comment]
never-tell-me-the-odds

**Comment by raybejjani** [comment]
never-tell-me-the-odds

File: pkg/kvstore/consul.go [hunk 1]
```diff
@@ -451,30 +451,40 @@ func (c *consulClient) Status() (string, error) {
...
+func (c *consulClient) DeleteIfLocked(ctx context.Context, key string, lock KVLocker) (err error) {
+	defer func() { Trace("DeleteIfLocked", err, logrus.Fields{fieldKey: key}) }()
+	return c.delete(ctx, key)
 }
 
 // Delete deletes a key
-func (c *consulClient) Delete(ctx context.Context, key string) error {
+func (c *consulClient) Delete(ctx context.Context, key string) (err error) {
+	defer func() { Trace("Delete", err, logrus.Fields{fieldKey: key}) }()
+	return c.Delete(ctx, key)
```
Comment (tgraf) [inline pkg/kvstore/consul.go]: Should this be `c.delete()`?

File: pkg/kvstore/consul.go [hunk 2]
```diff
@@ -483,12 +493,15 @@ func (c *consulClient) Delete(ctx context.Context, key string) error {
...
 }
 
 // Get returns value of key
-func (c *consulClient) Get(ctx context.Context, key string) ([]byte, error) {
+func (c *consulClient) Get(ctx context.Context, key string) (bv []byte, err error) {
+	defer func() { Trace("Get", err, logrus.Fields{fieldKey: key, fieldValue: string(bv)}) }()
+
 	duration := spanstat.Start()
 	qo := &consulAPI.QueryOptions{}
 	pair, _, err := c.KV().Get(key, qo.WithContext(ctx))
```
Comment (tgraf) [inline pkg/kvstore/consul.go:507]: `s/:=/=/`

File: pkg/kvstore/consul.go [hunk 3]
```diff
@@ -601,13 +651,17 @@ func (c *consulClient) CreateOnly(ctx context.Context, key string, value []byte,
...
 	// the presence of a conditional key if the key is not the key being
 	// manipulated
 	//
 	// Lock the conditional key to serialize all CreateIfExists() calls
 
+	defer func() {
+		Trace("CreateIfExists", err, logrus.Fields{fieldKey: key, fieldValue: string(value), fieldCondition: condKey, fieldAttachLease: lease})
+	}()
+
 	l, err := LockPath(ctx, condKey)
```
Comment (tgraf) [inline pkg/kvstore/consul.go]: You are converting other `:=` to `=` but not the ones in this function

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks good in general. I'm not sure we need a full-blown allocator for remote identities long-term but it's definitely fine as a fix to keep this minimal.

File: pkg/identity/cache/allocator.go [hunk 4]
```diff
@@ -437,5 +437,16 @@ func (m *CachingIdentityAllocator) ReleaseSlice(ctx context.Context, owner Ident
...
-	return m.IdentityAllocator.WatchRemoteKVStore(backend, m.identitiesPath)
+
+	remoteAllocatorBackend, err := kvstoreallocator.NewKVStoreBackend(m.identitiesPath, m.owner.GetNodeSuffix(), GlobalIdentity{}, backend)
+	if err != nil {
+		log.WithError(err).WithField(logfields.BackendName, backend).Error("Error setting up remote allocator backend")
+	}
+
+	remoteAlloc, err := allocator.NewAllocator(GlobalIdentity{}, remoteAllocatorBackend)
+	if err != nil {
+		log.WithError(err).Errorf("Unable to initialize remote Identity Allocator")
```
Comment (aanm) [inline pkg/identity/cache/allocator.go]: Shouldn't this return nil here?

File: pkg/identity/cache/allocator.go [hunk 5]
```diff
@@ -437,5 +437,16 @@ func (m *CachingIdentityAllocator) ReleaseSlice(ctx context.Context, owner Ident
 // syncs all identities to the local identity cache.
 func (m *CachingIdentityAllocator) WatchRemoteIdentities(backend kvstore.BackendOperations) *allocator.RemoteCache {
 	<-m.globalIdentityAllocatorInitialized
-	return m.IdentityAllocator.WatchRemoteKVStore(backend, m.identitiesPath)
+
+	remoteAllocatorBackend, err := kvstoreallocator.NewKVStoreBackend(m.identitiesPath, m.owner.GetNodeSuffix(), GlobalIdentity{}, backend)
+	if err != nil {
+		log.WithError(err).WithField(logfields.BackendName, backend).Error("Error setting up remote allocator backend")
```
Comment (aanm) [inline pkg/identity/cache/allocator.go]: Shouldn't this return nil here?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/identity/cache/allocator.go (hunk 5 above)
Comment (raybejjani) [reply pkg/identity/cache/allocator.go]: Yes, but it turned out to be worse. The caller of this didn't handle the nil case either. I made this return an error that is handled in the controller.

File: pkg/kvstore/consul.go (hunk 1 above)
Comment (raybejjani) [reply pkg/kvstore/consul.go]: yes it should! good catch

File: pkg/kvstore/consul.go (hunk 2 above)
Comment (raybejjani) [reply pkg/kvstore/consul.go:507]: `pair` is a new variable, though. If you're thinking that err won't be set correctly, I think that's only an issue with ifs (my test https://play.golang.org/p/QQSQpvv2C2P )

File: pkg/kvstore/consul.go (hunk 3 above)
Comment (raybejjani) [reply pkg/kvstore/consul.go]: oops. The trace should have been added to the public function just below this. I've switched it out. This function no longer needs the named err return variable.

**Comment by raybejjani** [comment]
never-tell-me-the-odds

**Comment by raybejjani** [comment]
never-tell-me-the-odds

**Comment by raybejjani** [comment]
never-tell-me-the-odds

**Review by tgraf — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Pull request merged**